include/mbedtls/fsio.h

```diff
@@ -0,0 +1,246 @@
+/**
+ * \file fsio.h
+ *
+ * \brief File read/write functions
+ *
+ *  Copyright (C) 2017, ARM Limited, All Rights Reserved
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  This file is part of mbed TLS (https://tls.mbed.org)
+ */
+#ifndef MBEDTLS_FILE_H
+#define MBEDTLS_FILE_H
+
+#if !defined(MBEDTLS_CONFIG_FILE)
+#include "config.h"
+#else
+#include MBEDTLS_CONFIG_FILE
+#endif
+
+#include <stddef.h>
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#if defined(MBEDTLS_FS_IO)
+/* If MBEDTLS_FS_IO is enabled then file IO functions should be made available
+ * via standard library or platform specific implementation. */
+
+
+/**
+ * Structure returned by mbedtls_stat().
+ *
+ * \note Future versions of the library may add more fields at the end of this
+ * 	 structure.
+ */
+typedef struct mbedtls_stat_tag
+{
+    uint16_t    type;   /** File type */
+/** A file that can be accessed as a stream, i.e. opened with mbedtls_fopen().
+ */
+#define MBEDTLS_FSIO_DT_FILE    0
+/** A directory that can be opened with mbedtls_opendir(). */
+#define MBEDTLS_FSIO_DT_DIR     1
+/** A directory entry that is recognized neither as a stream file nor as a
+ * directory. */
+#define MBEDTLS_FSIO_DT_OTHER   2
+} mbedtls_stat_t;
+
+/* Default and alternative implementation specific interfaces. */
+#if !defined(MBEDTLS_FS_IO_ALT)
+#include <stdio.h>
+#include <dirent.h>
+
+typedef FILE *  mbedtls_file_t;
```

Comment:
Running `make apidoc` omits a lot of documentation in this file because it shows the lack of Doxygen documentation for the macros, rather than the Doxygen documentation that is present for the functions of the same name. Please either duplicate the documentation (bleh) or arrange to run Doxygen with `MBEDTLS_FS_IO_ALT` set.

Reply:
I would need some help as I do not know how to do this.

Reply:
Looking back it isn't easy, because the documentation is generated from `scripts/config.pl realfull` which shouldn't include `MBEDTLS_FS_IO_ALT`. I have a pending patch to introduce a `docfull` config (internal reference: IOTSSL-1447), but I never got around to finishing it. In order to move on with this PR, which is important for on-target testing, I think it's ok to omit some functions in the documentation for now and track this as part of IOTSSL-1447.